be/src/storage/snapshot_manager.cpp

```diff
@@ -329,19 +329,28 @@ StatusOr<std::string> SnapshotManager::snapshot_incremental(const TabletSharedPt
     (void)FileUtils::remove_all(snapshot_dir);
     RETURN_IF_ERROR(FileUtils::create_dir(snapshot_dir));
 
-    // 3. Link files to snapshot directory.
+    // If tablet is PrimaryKey table, we should dump snapshot meta file first and then link files
+    // to snapshot directory
+    // The reason is ablet clone assumes rowset file is immutable, but during rowset apply for partial update,
+    // rowset file may be changed.
+    // When doing partial update, if dump snapshot meta file first, there are four conditions as below
+    //  1. rowset status is committed in meta, rowset file is partial rowset
+    //  2. rowset status is committed in meta, rowset file is partial rowset, but rowset is apply success after link file
```

Comment:
`but rowset is apply success after link file` what does this mean

Reply:
> `but rowset is apply success after link file` what does this mean

When we dump snapshot, `patial rowset` is committed but not apply, so we will do a hard link for partial rowset files. But when src BE start download snapshot, the`partial rowset` is applied.

This case should be the same with case1 theoretically. The partial rowset files will not be delete until snaoshot dir is deleted because of the hard link.  But to be on the safe side, it is differentiated from case1 and a test for case2 is added